plugin/evm/acp176/acp176.go

```diff
@@ -0,0 +1,197 @@
+	TargetToMax                   = 2    // multiplier to convert from target per second to max per second
+	TargetToPriceUpdateConversion = 43   // 43s ~= 30s * ln(2) which makes the price double at most every ~30 seconds
+	MaxTargetChangeRate           = 1024 // Controls the rate that the target can change per block.
+
+	targetToMaxCapacity = TargetToMax * TimeToFillCapacity
+	maxTargetExcess     = 1_024_950_627 // TargetConversion * ln(MaxUint64 / MinTargetPerSecond) + 1
+)
+
+var minBaseFee = big.NewInt(MinBaseFee)
+
+// State represents the current state of the gas pricing and constraints.
+type State struct {
+	Gas          gas.State
+	TargetExcess gas.Gas
+}
+
+// Target returns the target gas consumed per second.
+//
+// Target = MinTargetPerSecond * e^(TargetExcess / TargetConversion)
+func (s *State) Target() gas.Gas {
+	return gas.Gas(gas.CalculatePrice(
+		MinTargetPerSecond,
+		s.TargetExcess,
+		TargetConversion,
+	))
+}
+
+// MaxCapacity returns the maximum possible accrued gas capacity.
+func (s *State) MaxCapacity() gas.Gas {
+	targetPerSecond := s.Target()
+	maxCapacity, err := safemath.Mul(targetToMaxCapacity, targetPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	return maxCapacity
+}
+
+// BaseFee returns the current required fee per gas.
+//
+// BaseFee = MinBaseFee * e^(Excess / (Target() * TargetToPriceUpdateConversion))
+func (s *State) BaseFee() *big.Int {
+	target := s.Target()
+	priceUpdateConversion, err := safemath.Mul(TargetToPriceUpdateConversion, target)
+	if err != nil {
+		priceUpdateConversion = math.MaxUint64
+	}
+
+	bigExcess := new(big.Int).SetUint64(uint64(s.Gas.Excess))
+	bigPriceUpdateConversion := new(big.Int).SetUint64(uint64(priceUpdateConversion))
+	return fakeExponential(minBaseFee, bigExcess, bigPriceUpdateConversion)
+}
+
+// AdvanceTime increases the gas capacity and decreases the gas excess based on
+// the elapsed seconds.
+func (s *State) AdvanceTime(seconds uint64) {
+	targetPerSecond := s.Target()
+	maxPerSecond, err := safemath.Mul(TargetToMax, targetPerSecond)
+	if err != nil {
+		maxPerSecond = math.MaxUint64
+	}
+	maxCapacity, err := safemath.Mul(TimeToFillCapacity, maxPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	s.Gas = s.Gas.AdvanceTime(
+		maxCapacity,
+		maxPerSecond,
+		targetPerSecond,
+		seconds,
+	)
+}
+
+// ConsumeGas decreases the gas capacity and increases the gas excess by
+// gasUsed + extraGasUsed. If the gas capacity is insufficient, an error is
+// returned.
+func (s *State) ConsumeGas(
+	gasUsed uint64,
+	extraGasUsed *big.Int,
+) error {
+	var err error
+	s.Gas, err = s.Gas.ConsumeGas(gas.Gas(gasUsed))
+	if err != nil {
+		return err
+	}
+	if extraGasUsed != nil {
+		if !extraGasUsed.IsUint64() {
+			return fmt.Errorf("%w: extraGasUsed (%d) exceeds MaxUint64",
+				gas.ErrInsufficientCapacity,
+				extraGasUsed,
+			)
+		}
+		s.Gas, err = s.Gas.ConsumeGas(gas.Gas(extraGasUsed.Uint64()))
+	}
+	return err
```

Comment:
```suggestion
	if extraGasUsed == nil {
		return nil
	}
	if !extraGasUsed.IsUint64() {
		return fmt.Errorf("%w: extraGasUsed (%d) exceeds MaxUint64",
			gas.ErrInsufficientCapacity,
			extraGasUsed,
		)
	}
	s.Gas, err = s.Gas.ConsumeGas(gas.Gas(extraGasUsed.Uint64()))
	return err
```

Clearer if you reduce the nesting, especially the dangling `return err` that came from inside the `if`.